### PR TITLE
feat: monitor daily webhooks state cron

### LINKED ISSUE #15121: [CAL-3763] Trigger notification when daily webhook are not working
### Is your proposal related to a problem?

Trigger notification when daily webhooks state is invalid.

<sub>[CAL-3763](https://linear.app/calcom/issue/CAL-3763/trigger-notification-when-daily-webhook-are-not-working)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/15121
- Fixes CAL-3763 (Linear issue number - should be visible at the bottom of the GitHub issue description)

Screenshot 2024-06-13 at 10 40 51 PM

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] N/A I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

### CODE REVIEW

File: apps/web/pages/api/cron/monitorDailyWebhooks.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  if (process.env.CRON_API_KEY !== apiKey) {
+    res.status(401).json({ message: "Not authenticated" });
+    return;
+  }
+
+  const webhooks = await getWebhooks();
+
+  const inactiveWebhooks = webhooks.filter((webhook) => webhook.state !== "ACTIVE");
+  if (inactiveWebhooks.length) {
+    // Notify by sending email
```
Comment (Udit-takkar) [inline apps/web/pages/api/cron/monitorDailyWebhooks.ts:17]: TODO

**Comment by PeerRich** [comment]
is this being sent to users or to us admins?

File: .env.example [hunk 2]
```diff
@@ -367,3 +367,6 @@ RETELL_AI_KEY=
 
 # Used to disallow emails as being added as guests on bookings
 BLACKLISTED_GUEST_EMAILS=
+
+# Used to send alert email when daily webhooks goes down
+ALERT_EMAIL_TO=
```
Comment (hariombalhara) [inline .env.example:372]: If it is specific to daily webhook, we should be clear about it in the name.

File: packages/core/videoClient.ts [hunk 3]
```diff
@@ -384,6 +384,30 @@ const checkIfRoomNameMatchesInRecording = async (roomName: string, recordingId:
   return videoAdapter?.checkIfRoomNameMatchesInRecording?.(roomName, recordingId);
 };
 
+const getWebhooks = async () => {
```
Comment (hariombalhara) [inline packages/core/videoClient.ts:387]: Right now this is only working for daily. So, we should change it's name maybe for now atleast.

Ideally an app specific code shouldn't exist here.

File: apps/web/pages/api/cron/monitorDailyWebhooks.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    await sendSendgridMail(
+      {
+        to: process.env.ALERT_EMAIL_TO,
+        subject: "[HIGH] Alert: Daily Webhooks are not active",
+        html: `<body style="white-space: pre-wrap;">Daily Webhook url api/recorded-daily-video is not functional.</body>`,
```
Comment (hariombalhara) [inline apps/web/pages/api/cron/monitorDailyWebhooks.ts:24]: This is just communicating the problem. We should also inform about the possible ways to fix the problem

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Left some comments.

Also, a loom would help

**Comment by Udit-takkar** [comment]
> is this being sent to users or to us admins?
> because it is way too technical for users. and i wonder if we can track this as admins outside of our code base? 🤔
> also, why high priority?

When a recording is ready to be downloaded (after it has been stopped by user) we have a daily webhook (https://docs.daily.co/reference/rest-api/webhooks endpoint /api/recorded-daily-video ) that receives the payload with recording id, room name etc to be sent to email of attendees and trigger RECORDING_READY event. Now if this endpoint return a non success status code like 40x or 500 three times consecutively then the daily webhooks goes to INVALID state and doesn't trigger /api/recorded-daily-video enpoint unless we restart it again.

File: .env.example (hunk 2 above)
Comment (Udit-takkar) [reply .env.example:372]: i think we can use this variable in the future for any other thing to send trigger notification to use

Comment (Udit-takkar) [reply .env.example]: Ok for now i have changed the name `DAILY_WEBHOOKS_ALERT_EMAIL_TO`

File: apps/web/pages/api/cron/monitorDailyWebhooks.ts (hunk 4 above)
Comment (Udit-takkar) [reply apps/web/pages/api/cron/monitorDailyWebhooks.ts:24]: Actually we have debug this issue ourselves. there is no info about the error

Comment (Udit-takkar) [reply apps/web/pages/api/cron/monitorDailyWebhooks.ts:24]: more info here https://github.com/calcom/cal.com/pull/15126#issuecomment-2167439247

**Comment by hariombalhara** [comment]
> > is this being sent to users or to us admins?
> > because it is way too technical for users. and i wonder if we can track this as admins outside of our code base? 🤔
> > also, why high priority?
> 
> When a recording is ready to be downloaded (after it has been stopped by user) we have a daily webhook ([docs.daily.co/reference/rest-api/webhooks](https://docs.daily.co/reference/rest-api/webhooks) endpoint /api/recorded-daily-video ) that receives the payload with recording id, room name etc to be sent to email of attendees and trigger RECORDING_READY event.
> 
> Now if this endpoint return a non success status code like 40x or 500 three times consecutively then the daily webhooks goes to FAILED state and doesn't trigger /api/recorded-daily-video enpoint unless we restart it again.

So, it affects all the users using the Daily location after that? 

Also, not sure if it is discussed but in addition to this email we should restart it automatically(if possible).

File: packages/core/getDailyWebhooks.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    dailyAppKeys = await getDailyAppKeys();
+  } catch (e) {
+    console.error("Error: Cal video provider is not installed.");
+    return;
+  }
+  const [videoAdapter] = await getVideoAdapters([
+    {
+      id: 0,
+      appId: "daily-video",
+      type: "daily_video",
```
Comment (hariombalhara) [inline packages/core/getDailyWebhooks.ts:17]: how about importing _metadata here and use the values from there.

File: packages/types/VideoApiAdapter.d.ts [hunk 6]
```diff
@@ -41,6 +42,8 @@ export type VideoApiAdapter =
       ): Promise<TGetTranscriptAccessLink["transcription"] | { message: string }>;
 
       checkIfRoomNameMatchesInRecording?(roomName: string, recordingId: string): Promise<boolean>;
+
+      getWebhooks?(): Promise<TGetDailyWebhooks>;
```
Comment (hariombalhara) [inline packages/types/VideoApiAdapter.d.ts:46]: It should be a generic schema like `TGetWebhook` as this is a type specification for all apps.

**Review by hariombalhara — APPROVED** [review]
Code looks good except for the changes I suggested. Didn't test this.

**Comment by Udit-takkar** [comment]
I have discussed an alternate solution with daily team.  I'll update the retry type to "exponential"

**Pull request closed without merging**